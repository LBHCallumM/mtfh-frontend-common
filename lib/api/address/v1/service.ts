import { CommonAuth } from "../../../auth";
import { config } from "../../../config";
import { AxiosSWRConfiguration, getAxiosInstance, useAxiosSWR } from "../../../http";

import type { Address } from "./types";

export interface AddressAPIResponse {
  data: { address: Address[] };
}

interface SearchAddressResponse {
  addresses?: Address[];
  error?: { code: number };
}

export const searchAddress = async (
  postCode: string,
  auth: CommonAuth,
): Promise<SearchAddressResponse> => {
  const axiosInstance = getAxiosInstance(auth);

  return axiosInstance
    .get<AddressAPIResponse>(`${config.addressApiUrlV1}/addresses?postcode=${postCode}`, {
      headers: {
        "skip-x-correlation-id": true,
      },
    })
    .then((res) => ({ addresses: res.data.data.address }))
    .catch((res) => {
      if (res.message.toLowerCase().indexOf("network") !== -1) {
        return { error: { code: 500 } };
      }
      return res;
    });
};

export const getAddressViaUprn = async (
  uprn: string,
  auth: CommonAuth,
): Promise<SearchAddressResponse> => {
  const axiosInstance = getAxiosInstance(auth);

  return axiosInstance
    .get<AddressAPIResponse>(`${config.addressApiUrlV1}/addresses?uprn=${uprn}`, {
      headers: {
        "skip-x-correlation-id": true,
      },
    })
    .then((res) => ({ addresses: res.data.data.address }))
    .catch((res) => {
      if (res.message.toLowerCase().indexOf("network") !== -1) {
        return { error: { code: 500 } };
      }
      return res;
    });
};

export const useAddressLookup = (
  auth: CommonAuth,
  postCode?: string | null,
  options: AxiosSWRConfiguration<AddressAPIResponse> = {},
) => {
  return useAxiosSWR<AddressAPIResponse>(
    postCode ? `${config.addressApiUrlV1}/addresses?postcode=${postCode}` : null,
    auth,
    {
      ...options,
      timeout: 5000,
      headers: {
        ...options.headers,
        "skip-x-correlation-id": true,
      },
    },
  );
};
