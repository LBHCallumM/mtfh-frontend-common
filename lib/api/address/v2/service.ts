import { config } from "../../../config";
import type { Address } from "./types";

interface AddressAPIv2Response {
  data: { address: Address[] };
}

interface SearchAddressResponse {
  addresses?: Address[];
  error?: { code: number };
}

export const getAddressViaUprn = (UPRN: string): Promise<SearchAddressResponse> =>
  axiosInstance
    .get<AddressAPIv2Response>(`${config.addressApiUrlV2}/addresses/${UPRN}`, {
      headers: {
        "skip-x-correlation-id": true,
      },
    })
    .then((res) => ({ addresses: res.data.data.address }))
    .catch((res) => {
      if (res.message.toLowerCase().indexOf("network") !== -1) {
        return { error: { code: 500 } };
      }
      return res;
    });

export const useAddressLookupUprn = (
  uprn?: string | null,
  options: AxiosSWRConfiguration<AddressAPIv2Response> = {},
) => {
  return useAxiosSWR<AddressAPIv2Response>(
    uprn ? `${config.addressApiUrlV2}/addresses/${uprn}` : null,
    {
      ...options,
      timeout: 5000,
      headers: {
        ...options.headers,
        "skip-x-correlation-id": true,
      },
    },
  );
};
