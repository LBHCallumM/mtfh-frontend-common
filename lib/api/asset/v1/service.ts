import { CommonAuth } from "../../../auth";
import { config } from "../../../config";
import {
  AxiosSWRConfiguration,
  AxiosSWRResponse,
  getAxiosInstance,
  useAxiosSWR,
} from "../../../http";
import { Asset, EditAssetAddressRequest } from "./types";

export const useAsset = (
  id: string | null,
  auth: CommonAuth,
  options?: AxiosSWRConfiguration<Asset>,
): AxiosSWRResponse<Asset> => {
  return useAxiosSWR(id && `${config.assetApiUrlV1}/assets/${id}`, auth, options);
};

export const patchAsset = async (
  id: string,
  assetAddress: EditAssetAddressRequest,
  assetVersion: string,
  auth: CommonAuth,
): Promise<void> => {
  return new Promise<void>((resolve, reject) => {
    const axiosInstance = getAxiosInstance(auth);

    axiosInstance
      .patch(`${config.assetApiUrlV1}/assets/${id}/address`, assetAddress, {
        headers: {
          "If-Match": assetVersion,
        },
      })
      .then(() => resolve())
      .catch(() => reject());
  });
};
