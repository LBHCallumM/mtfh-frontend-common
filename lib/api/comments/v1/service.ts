import { stringify } from "query-string";

import { CommonAuth } from "../../../auth";
import { config } from "../../../config";
import {
  AxiosSWRInfiniteResponse,
  getAxiosInstance,
  useAxiosSWRInfinite,
} from "../../../http";

import type { Comment } from "./types";

export interface GetCommentsByTargetIdResponse {
  results: Comment[];
  paginationDetails: {
    nextToken: string | null;
  };
}

export interface GetCommentsByIdRequestData {
  targetId: string;
  pageSize?: number;
  paginationToken?: string | null;
}

export const useComments = (
  id: string,
  auth: CommonAuth,
  pageSize = 5,
): AxiosSWRInfiniteResponse<GetCommentsByTargetIdResponse> => {
  return useAxiosSWRInfinite<GetCommentsByTargetIdResponse>((page, previous) => {
    if (previous && !previous?.paginationDetails?.nextToken) {
      return null;
    }

    const params: GetCommentsByIdRequestData = {
      targetId: id,
      pageSize,
    };

    if (page !== 0 && previous?.paginationDetails.nextToken) {
      params.paginationToken = previous.paginationDetails.nextToken;
    }

    return `${config.notesApiUrlV1}/notes?${stringify(params)}`;
  }, auth);
};

export type PostCommentRequestData = Omit<
  Comment,
  "id" | "categorisation" | "author" | "createdAt"
>;

export const addComment = async (
  data: PostCommentRequestData,
  auth: CommonAuth,
): Promise<Comment> => {
  const axiosInstance = getAxiosInstance(auth);

  const { sub: id, email, name: fullName } = auth.user;

  const { data: comment } = await axiosInstance.post(`${config.notesApiUrlV1}/notes`, {
    ...data,
    createdAt: new Date().toISOString(),
    author: {
      id,
      email,
      fullName,
    },
  });
  return comment;
};
