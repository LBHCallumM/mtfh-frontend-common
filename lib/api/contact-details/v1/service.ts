import { config } from "../../../config";
import { getAxiosInstance } from "../../../http";

export const deleteContactDetail = async (
  id: string,
  targetId: string,
): Promise<void> => {
  const axiosInstance = getAxiosInstance();

  const response = await axiosInstance.delete(
    `${config.contactDetailsApiUrlV1}/contactDetails?id=${id}&targetId=${targetId}`,
  );
  return response.data;
};
