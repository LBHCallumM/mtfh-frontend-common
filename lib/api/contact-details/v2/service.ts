import { CommonAuth } from "../../../auth";
import { config } from "../../../config";
import {
  AxiosSWRConfiguration,
  AxiosSWRResponse,
  getAxiosInstance,
  useAxiosSWR,
} from "../../../http";
import { removeWhitespace } from "../../../utils";
import {
  ContactDetail,
  ContactDetailsAddressTypes,
  ContactDetailsPhoneTypes,
  ContactInformation,
  ContactInformationContactTypes,
} from "./types";

export interface ContactDetailsResponse {
  results: ContactDetail[];
}

export const useContactDetails = (
  id: string,
  auth: CommonAuth,
  options?: AxiosSWRConfiguration<ContactDetailsResponse>,
): AxiosSWRResponse<ContactDetailsResponse> =>
  useAxiosSWR(
    id && `${config.contactDetailsApiUrlV2}/contactDetails?targetId=${id}`,
    auth,
    options,
  );

export const addContactDetail = async (
  id: string,
  data: ContactInformation,
  auth: CommonAuth,
): Promise<ContactDetail> => {
  const axiosInstance = getAxiosInstance(auth);

  const response = await axiosInstance.post(
    `${config.contactDetailsApiUrlV2}/contactDetails`,
    {
      targetId: id,
      targetType: "person",
      contactInformation: {
        ...data,
        value: removeWhitespace(data.value),
      },
      sourceServiceArea: {
        area: "Housing",
        isDefault: true,
      },
    },
  );

  return response.data;
};

export const addEmailContact = async (
  auth: CommonAuth,
  id: string,
  email: string,
  description?: string,
): Promise<ContactDetail> =>
  addContactDetail(
    id,
    {
      contactType: ContactInformationContactTypes.EMAIL,
      value: email,
      description,
    },
    auth,
  );

export const addPhoneContact = async (
  auth: CommonAuth,
  id: string,
  phone: string,
  type: ContactDetailsPhoneTypes,
  description?: string,
): Promise<ContactDetail> =>
  addContactDetail(
    id,
    {
      contactType: ContactInformationContactTypes.PHONE,
      value: phone,
      subType: type,
      description,
    },
    auth,
  );

export interface AddCorrespondenceAddressArgs {
  id: string;
  addressLine1: string;
  addressLine2?: string | null;
  addressLine3?: string | null;
  addressLine4?: string | null;
  postCode: string;
  description?: string | null;
  isOverseasAddress?: boolean;
  overseasAddress?: string | null;
}

export const addCorrespondenceAddress = async (
  {
    id,
    addressLine1,
    addressLine2 = null,
    addressLine3 = null,
    addressLine4 = null,
    postCode,
    description = null,
    isOverseasAddress = false,
    overseasAddress = null,
  }: AddCorrespondenceAddressArgs,
  auth: CommonAuth,
): Promise<ContactDetail> =>
  addContactDetail(
    id,
    {
      contactType: ContactInformationContactTypes.ADDRESS,
      value: "",
      subType: ContactDetailsAddressTypes.CORRESPONDENCE_ADDRESS,
      description,
      addressExtended: {
        uprn: null,
        addressLine1,
        addressLine2,
        addressLine3,
        addressLine4,
        postCode,
        overseasAddress,
        isOverseasAddress,
      },
    },
    auth,
  );
