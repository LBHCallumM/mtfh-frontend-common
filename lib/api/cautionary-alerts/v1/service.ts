import { TenureSummary } from "../../../api/person/v1";
import { CommonAuth } from "../../../auth";

import { config } from "../../../config";
import {
  AxiosSWRConfiguration,
  AxiosSWRResponse,
  getAxiosInstance,
  useAxiosSWR,
} from "../../../http";
import { Alert, CautionaryAlert } from "./types";

export const useCautionaryAlert = (
  id: string | null,
  auth: CommonAuth,
  options?: AxiosSWRConfiguration<any>,
): AxiosSWRResponse<CautionaryAlert> => {
  return useAxiosSWR(
    id && `${config.cautionaryApiUrlV1}/cautionary-alerts/persons/${id}`,
    auth,
    options,
  );
};

export const usePropertyCautionaryAlert = (
  propertyRef: string | null,
  auth: CommonAuth,
  options?: AxiosSWRConfiguration<any>,
): AxiosSWRResponse<CautionaryAlert> => {
  return useAxiosSWR(
    propertyRef &&
      `${config.cautionaryApiUrlV1}/cautionary-alerts/properties-new/${propertyRef}`,
    auth,
    options,
  );
};

export type PostCautionaryAlertRequestData = Pick<Alert, "assureReference"> & {
  incidentDate: string;
  incidentDescription: string;
  alert: {
    code: string;
    description: string;
  };
  assetDetails?: Pick<TenureSummary, "propertyReference" | "uprn"> & {
    id: string;
    fullAddress: string;
  };
  personDetails: {
    id: string;
    name: string;
  };
};

export const addCautionaryAlert = async (
  data: PostCautionaryAlertRequestData,
  auth: CommonAuth,
): Promise<Alert> => {
  const axiosInstance = getAxiosInstance(auth);

  const { data: alert } = await axiosInstance.post(
    `${config.cautionaryApiUrlV1}/cautionary-alerts/`,
    data,
  );
  return {
    ...alert,
    alertCode: alert.code,
    description: alert.cautionOnSystem,
    personName: alert.name,
    startDate: alert.dateOfIncident,
  };
};
