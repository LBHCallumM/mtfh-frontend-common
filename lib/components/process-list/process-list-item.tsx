import React, { useMemo } from "react";
import { Link as RouterLink } from "react-router-dom";

import { useTenure } from "@mtfh/common/lib/api/tenure/v1";
import { formatDate } from "@mtfh/common/lib/utils";

import { Link } from "../link";

import "./process-list-item.scss";

export interface ProcessListItemParameters {
  processId: string;
  processType: string;
  tenureId: string;
  tenantId?: string;
  fullName?: string;
  createdAt: string;
  processName: string;
  status: string;
}

export const ProcessListItem = (props: ProcessListItemParameters): JSX.Element => {
  const {
    processType,
    tenantId,
    fullName,
    createdAt,
    processName,
    status,
    processId,
    tenureId,
  } = props;

  const { error: tenureError, data: tenure } = useTenure(tenureId);

  const address = tenure?.tenuredAsset?.fullAddress;

  const createdAtStr = useMemo(() => formatDate(createdAt), [createdAt]);

  return (
    <div className="process">
      {!tenureError && (
        <>
          <div className="process__item">
            <div>
              <Link
                as={RouterLink}
                to={`/person/${tenantId}`}
                variant="link"
                target="_blank"
                className="lbh-heading-h4"
              >
                {fullName}
              </Link>
            </div>
            <div style={{ marginTop: 0 }}>
              <Link
                as={RouterLink}
                to={`/person/${tenantId}`}
                variant="link"
                target="_blank"
              >
                {address}
              </Link>
            </div>
          </div>
          <div className="process__item --processName">
            <Link
              as={RouterLink}
              to={`/processes/${processType}/${processId}`}
              variant="link"
              target="_blank"
              className="lbh-heading-h4"
            >
              {processName}
            </Link>
          </div>
          <div className="process__item --createdAt">
            <div className="process__title">Initiated</div>
            <div className="process__createdAt">{createdAtStr}</div>
          </div>
          <div className="process__item --status">
            <span className="govuk-tag lbh-tag">{status}</span>
          </div>
        </>
      )}
    </div>
  );
};
