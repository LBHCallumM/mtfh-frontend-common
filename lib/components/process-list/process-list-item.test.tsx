import React from "react";

import { mockActiveTenureV1, render } from "@hackney/mtfh-test-utils";

import { Tenure } from "@mtfh/common/lib/api/tenure/v1";
import { AxiosSWRResponse } from "@mtfh/common/lib/http";

import * as tenureV1 from "../../api/tenure/v1/service";
import { ProcessListItem, ProcessListItemParameters } from "./process-list-item";

describe("ProcessListItem", () => {
  test("ProcessListItem renders correctly", () => {
    jest.spyOn(tenureV1, "useTenure").mockReturnValue({
      data: mockActiveTenureV1,
    } as AxiosSWRResponse<Tenure>);

    const processListItemParams: ProcessListItemParameters = {
      processId: "39273cf3-6a78-45e1-32d4-b9637e310527",
      processType: "soletojoint",
      tenureId: "39273cf3-6a78-45e1-32d4-b9637e310527",
      tenantId: "39273cf3-6a78-45e1-32d4-b9637e310527",
      fullName: "Mr Ersan Kuneri",
      createdAt: "2022-06-24T07:47:21.580066Z",
      processName: "Sole to Joint",
      status: "Process Closed",
    };

    const { container } = render(<ProcessListItem {...processListItemParams} />);

    expect(container).toMatchSnapshot();
  });
});
