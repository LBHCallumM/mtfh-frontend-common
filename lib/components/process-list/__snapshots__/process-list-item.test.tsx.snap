// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessListItem ProcessListItem renders correctly 1`] = `
<div>
  <div
    class="process"
  >
    <div
      class="process__item"
    >
      <div>
        <a
          class="govuk-link lbh-link lbh-link--no-visited-state lbh-heading-h4"
          href="/person/39273cf3-6a78-45e1-32d4-b9637e310527"
          target="_blank"
        >
          Mr Ersan Kuneri
        </a>
      </div>
      <div
        style="margin-top: 0px;"
      >
        <a
          class="govuk-link lbh-link lbh-link--no-visited-state"
          href="/person/39273cf3-6a78-45e1-32d4-b9637e310527"
          target="_blank"
        >
          72251 Champlin Trafficway, 99647
        </a>
      </div>
    </div>
    <div
      class="process__item --processName"
    >
      <a
        class="govuk-link lbh-link lbh-link--no-visited-state lbh-heading-h4"
        href="/processes/soletojoint/39273cf3-6a78-45e1-32d4-b9637e310527"
        target="_blank"
      >
        Sole to Joint
      </a>
    </div>
    <div
      class="process__item --createdAt"
    >
      <div
        class="process__title"
      >
        Initiated
      </div>
      <div
        class="process__createdAt"
      >
        24/06/2022
      </div>
    </div>
    <div
      class="process__item --status"
    >
      <span
        class="govuk-tag lbh-tag"
      >
        Process Closed
      </span>
    </div>
  </div>
</div>
`;
